hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomSelfDestructOperation.java

```diff
@@ -92,65 +97,135 @@ public OperationResult execute(@NonNull final MessageFrame frame, @NonNull final
             final var tbdAddress = frame.getRecipientAddress();
             final var proxyWorldUpdater = (ProxyWorldUpdater) frame.getWorldUpdater();
 
-            // Now proceed with the self-destruct
+            // In EVM the SELFDESTRUCT operation never fails.  It always sweeps ETH, and the contract
+            // is either deleted or not (per EIP-6780).
+            //
+            // In Hedera we have to allow for our semantics for transfers.  Notably, we can't
+            // transfer hbar unless the signature requirements are met on the transaction, we can't
+            // burn hbar, and we don't allow deletion of a contract which is a token treasury.
+            // There's also a restriction due to our performance guarantees: We don't allow contracts
+            // holding native tokens to self destruct because all the tokens would have to be
+            // transferred in the current `handleTransaction` call and if there were too many tokens
+            // it would be too expensive (in CPU/memory/database resources) to transfer them all.
+            //
+            // If the beneficiary account is the contract itself then we have two cases:
```

Comment:
Should there be bdd tests for these conditions?

Reply:
Provided now

Reply:
are they in a different pr?